diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/CHANGES ./RegMiner4APR-Regression-Bugs/BIC/CHANGES
--- ./RegMiner4APR-Regression-Bugs/WORKING/CHANGES	2024-09-08 01:36:15.569818784 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/CHANGES	2024-09-08 01:36:15.561818727 +1000
@@ -32,6 +32,9 @@
     the one that defines a character set.
     <https://github.com/jhy/jsoup/pull/835>
 
+  * Bugfix: when parsing unknown tags in case-sensitive HTML mode, end tags would not close scope correctly.
+    <https://github.com/jhy/jsoup/issues/819>
+
 *** Release 1.10.2 [2017-Jan-02]
  * Improved startup time, particularly on Android, by reducing garbage generation and CPU execution time when loading
    the HTML entity files. About 1.72x faster in this area.
diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/jsoup/parser/HtmlTreeBuilderState.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/jsoup/parser/HtmlTreeBuilderState.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/jsoup/parser/HtmlTreeBuilderState.java	2024-09-08 01:36:15.569818784 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/jsoup/parser/HtmlTreeBuilderState.java	2024-09-08 01:36:15.561818727 +1000
@@ -761,7 +761,7 @@
         }
 
         boolean anyOtherEndTag(Token t, HtmlTreeBuilder tb) {
-            String name = t.asEndTag().normalName();
+            String name = t.asEndTag().name(); // matches with case sensitivity if enabled
             ArrayList<Element> stack = tb.getStack();
             for (int pos = stack.size() -1; pos >= 0; pos--) {
                 Element node = stack.get(pos);
