diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/github/javafaker/Number.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/github/javafaker/Number.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/github/javafaker/Number.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/github/javafaker/Number.java
@@ -39,7 +39,7 @@
      * @param max exclusive (unless min == max)
      */
     public long numberBetween(long min, long max) {
-        return decimalBetween(min,max).longValue();
+        return decimalBetween(min,max).setScale(0, BigDecimal.ROUND_HALF_DOWN).longValue();
     }
     
     /**