diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/googlecode/javaewah/EWAHCompressedBitmap.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/googlecode/javaewah/EWAHCompressedBitmap.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/googlecode/javaewah/EWAHCompressedBitmap.java	2026-01-02 02:44:09.837878765 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/googlecode/javaewah/EWAHCompressedBitmap.java	2026-01-02 02:41:38.773960344 +1100
@@ -298,6 +298,11 @@
     @Override
     public void addStreamOfLiteralWords(final Buffer buffer, final int start,
                                         final int number) {
+        if((start + 1 == buffer.sizeInWords()) && (buffer.getWord(start) == 0)) {
+            insertEmptyWord(false);
+            return;
+        }
         int leftOverNumber = number;
         while (leftOverNumber > 0) {
             final int numberOfLiteralWords = this.rlw.getNumberOfLiteralWords();
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/googlecode/javaewah32/EWAHCompressedBitmap32.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/googlecode/javaewah32/EWAHCompressedBitmap32.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/googlecode/javaewah32/EWAHCompressedBitmap32.java	2026-01-02 02:44:09.841878764 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/googlecode/javaewah32/EWAHCompressedBitmap32.java	2026-01-02 02:41:38.777960342 +1100
@@ -274,6 +274,11 @@
     @Override
     public void addStreamOfLiteralWords(final Buffer32 buffer, final int start,
                                         final int number) {
+        if((start + 1 == buffer.sizeInWords()) && (buffer.getWord(start) == 0)) {
+            insertEmptyWord(false);
+            return;
+        }
         int leftOverNumber = number;
         while (leftOverNumber > 0) {
             final int numberOfLiteralWords = this.rlw.getNumberOfLiteralWords();
