diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/apache/commons/beanutils/converters/AbstractConverter.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/apache/commons/beanutils/converters/AbstractConverter.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/apache/commons/beanutils/converters/AbstractConverter.java	2025-12-14 20:03:55.293024097 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/apache/commons/beanutils/converters/AbstractConverter.java	2025-12-15 02:34:29.651433208 +1100
@@ -43,11 +44,18 @@
  * @version $Id$
  * @since 1.8.0
  */
-public abstract class AbstractConverter implements Converter {
+public abstract class AbstractConverter<D> implements Converter {
 
     /** Debug logging message to indicate default value configuration */
     private static final String DEFAULT_CONFIG_MSG =
@@ -71,7 +79,7 @@
     /**
      * The default value specified to our Constructor, if any.
      */
-    private Object defaultValue = null;
+    private D defaultValue = null;
 
     // ----------------------------------------------------------- Constructors
 
@@ -90,7 +98,7 @@
      * if the value to be converted is missing or an error
      * occurs converting the value.
      */
-    public AbstractConverter(Object defaultValue) {
+    public AbstractConverter(D defaultValue) {
         setDefaultValue(defaultValue);
     }
 
@@ -112,16 +120,21 @@
      * @throws ConversionException if conversion cannot be performed
      * successfully and no default is specified.
      */
-    public Object convert(Class type, Object value) {
+    public <T> T convert(Class<T> type, Object value) {
 
-        Class sourceType  = value == null ? null : value.getClass();
-        Class targetType  = primitive(type  == null ? getDefaultType() : type);
+        if (type == null) {
+            return convertToDefaultType(value);
+        }
+
+        Class<?> sourceType  = value == null ? null : value.getClass();
+        Class<T> targetType  = primitive(type);
 
         if (log().isDebugEnabled()) {
             log().debug("Converting"
@@ -141,7 +154,7 @@
         try {
             // Convert --> String
             if (targetType.equals(String.class)) {
-                return convertToString(value);
+                return targetType.cast(convertToString(value));
 
             // No conversion necessary
             } else if (targetType.equals(sourceType)) {
@@ -149,7 +162,7 @@
                     log().debug("    No conversion required, value is already a "
                                     + toString(targetType));
                 }
-                return value;
+                return targetType.cast(value);
 
             // Convert --> Type
             } else {
@@ -158,7 +171,7 @@
                     log().debug("    Converted to " + toString(targetType) +
                                    " value '" + result + "'");
                 }
-                return result;
+                return targetType.cast(result);
             }
         } catch (Throwable t) {
             return handleError(targetType, value, t);
@@ -194,7 +207,7 @@
      * @return The converted value.
      * @throws Throwable if an error occurs converting to the specified type
      */
-    protected abstract Object convertToType(Class type, Object value) throws Throwable;
+    protected abstract <T> T convertToType(Class<T> type, Object value) throws Throwable;
 
     /**
      * Return the first element from an Array (or Collection)
@@ -218,7 +231,7 @@
             }
         }
         if (value instanceof Collection) {
-            Collection collection = (Collection)value;
+            Collection<?> collection = (Collection<?>)value;
             if (collection.size() > 0) {
                 return collection.iterator().next();
             } else {
@@ -241,7 +254,7 @@
      * @throws ConversionException if no default value has been
      * specified for this {@link Converter}.
      */
-    protected Object handleError(Class type, Object value, Throwable cause) {
+    protected <T> T handleError(Class<T> type, Object value, Throwable cause) {
         if (log().isDebugEnabled()) {
             if (cause instanceof ConversionException) {
                 log().debug("    Conversion threw ConversionException: " + cause.getMessage());
@@ -279,15 +292,16 @@
      * @throws ConversionException if no default value has been
      * specified for this {@link Converter}.
      */
-    protected Object handleMissing(Class type) {
+    protected <T> T handleMissing(Class<T> type) {
 
         if (useDefault || type.equals(String.class)) {
             Object value = getDefault(type);
@@ -295,8 +309,8 @@
                 try {
                     value = convertToType(type, defaultValue);
                 } catch (Throwable t) {
-                    log().error("    Default conversion to " + toString(type)
-                            + "failed: " + t);
+                    throw new ConversionException("Default conversion to " + toString(type)
+                            + " failed.", t);
                 }
             }
             if (log().isDebugEnabled()) {
@@ -304,7 +318,8 @@
                         + (value == null ? "" : toString(value.getClass()) + " ")
                         + "value '" + defaultValue + "'");
             }
-            return value;
+            return type.cast(value);
         }
 
         ConversionException cex =  new ConversionException("No value specified for '" +
@@ -348,7 +363,7 @@
      *
      * @return The default type this <code>Converter</code> handles.
      */
-    protected abstract Class getDefaultType();
+    protected abstract Class<? extends D> getDefaultType();
 
     /**
      * Return the default value for conversions to the specified
@@ -356,10 +371,10 @@
      * @param type Data type to which this value should be converted.
      * @return The default value for the specified type.
      */
-    protected Object getDefault(Class type) {
+    protected Object getDefault(Class<?> type) {
         if (type.equals(String.class)) {
             return null;
        } else  {
@@ -394,31 +409,34 @@
      * @param type The class type to check.
      * @return The converted type.
      */
-     Class primitive(Class type) {
+    @SuppressWarnings("unchecked")
+    <T> Class<T> primitive(Class<T> type) {
         if (type == null || !type.isPrimitive()) {
             return type;
         }
