diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/release-notes/VERSION-2.x ./RegMiner4APR-Regression-Bugs/BIC/release-notes/VERSION-2.x
--- ./RegMiner4APR-Regression-Bugs/WORKING/release-notes/VERSION-2.x	2025-12-11 14:57:18.037109157 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/release-notes/VERSION-2.x	2025-12-11 15:01:15.529296279 +1100
@@ -31,6 +31,8 @@
   for case of q2 == null
 #692: UTF32Reader ArrayIndexOutOfBoundsException
  (reported by Fabian M)
+#694: Improve exception/JsonLocation handling for binary content: don't
+  show content, include byte offset
 
 2.12.3 (not yet released)
 2.12.2 (03-Mar-2021)
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/fasterxml/jackson/core/io/ContentReference.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/fasterxml/jackson/core/io/ContentReference.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/fasterxml/jackson/core/io/ContentReference.java	2025-12-11 14:57:15.769107408 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/fasterxml/jackson/core/io/ContentReference.java	2025-12-11 15:01:12.581293918 +1100
@@ -24,6 +24,9 @@
     /**
      * Constant that may be used when source/target content is not known
      * (or not exposed).
+     *<p>
+     * NOTE: As of 2.13 assume to contain Binary content, meaning that no
+     * content snippets will be included.
      */
     protected final static ContentReference UNKNOWN_CONTENT =
             new ContentReference(false, null);
@@ -107,17 +110,25 @@
      * content about which only minimal amount of information is available.
      * Assumed not to contain textual content (no snippet displayed).
      *
+     * @param isContentTextual Is raw content assumed to have textual content
+     *    ({@code true}) or binary ({@code false})
      * @param rawContent Underlying raw content access
      *
      * @return Instance with minimal information about content (basically just
      *    raw content reference without offsets
      */
-    public static ContentReference rawReference(Object rawContent) {
+    public static ContentReference rawReference(boolean isContentTextual,
+            Object rawContent)
+    {
         // Just to avoid russian-doll-nesting, let's:
         if (rawContent instanceof ContentReference) {
             return (ContentReference) rawContent;
         }
-        return new ContentReference(false, rawContent);
+        return new ContentReference(isContentTextual, rawContent);
+    }
+
+    public static ContentReference rawReference(Object rawContent) {
+        return rawReference(false, rawContent);
     }
 
     /*
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/fasterxml/jackson/core/JsonLocation.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/fasterxml/jackson/core/JsonLocation.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/fasterxml/jackson/core/JsonLocation.java	2025-12-11 14:57:15.765107406 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/fasterxml/jackson/core/JsonLocation.java	2025-12-11 15:01:12.573293912 +1100
@@ -151,11 +151,48 @@
         return appendOffsetDescription(new StringBuilder(40)).toString();
     }
 
-    public StringBuilder appendOffsetDescription(StringBuilder sb) {
-        sb.append("line: ");
-        sb.append(_lineNr);
-        sb.append(", column: ");
-        sb.append(_columnNr);
+    // @since 2.13
+    public StringBuilder appendOffsetDescription(StringBuilder sb)
+    {
+        // 04-Apr-2021, tatu: [core#694] For binary content, we have no line
+        //    number or column position indicators; try using what we do have
+        //    (if anything)
+
+        if (_contentReference.hasTextualContent()) {
+            sb.append("line: ");
+            // should be 1-based, but consider -1 to be canonical "got none"
+            if (_lineNr >= 0) {
+                sb.append(_lineNr);
+            } else {
+                sb.append("UNKNOWN");
+            }
+            sb.append(", column: ");
+            if (_columnNr >= 0) { // same here
+                sb.append(_columnNr);
+            } else {
+                sb.append("UNKNOWN");
+            }
+        } else {
+            // 04-Apr-2021, tatu: Ideally byte formats would not need line/column
+            //    info, but for backwards-compatibility purposes (Jackson 2.x),
+            //    will leave logic here
+            if (_lineNr > 0) { // yes, require 1-based in case of allegedly binary content
+                sb.append("line: ").append(_lineNr);
+                if (_columnNr > 0) {
+                    sb.append(", column: ");
+                    sb.append(_columnNr);
+                }
+            } else {
+                sb.append("byte offset: #");
+                // For binary formats, total bytes should be the canonical offset
+                // for token/current location
+                if (_totalBytes >= 0) {
+                    sb.append(_totalBytes);
+                } else {
+                    sb.append("UNKNOWN");
+                }
+            }
+        }
         return sb;
     }
 
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/com/fasterxml/jackson/core/TestLocation.java ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/com/fasterxml/jackson/core/TestLocation.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/com/fasterxml/jackson/core/TestLocation.java	2025-12-11 14:57:15.749107394 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/com/fasterxml/jackson/core/TestLocation.java	2025-12-11 15:01:12.553293895 +1100
@@ -28,8 +28,8 @@
 
     public void testBasicToString()
     {
-        // no location:
-        assertEquals("[Source: UNKNOWN; line: 3, column: 2]",
+        // no location; presumed to be Binary due to defaulting
+        assertEquals("[Source: UNKNOWN; byte offset: #10]",
                 new JsonLocation(null, 10L, 10L, 3, 2).toString());
 
         // Short String
@@ -51,12 +51,12 @@
 
         // Class<?> that specifies source type
         assertEquals("[Source: (InputStream); line: 1, column: 2]",
-                new JsonLocation(_rawSourceRef(InputStream.class), 10L, 10L, 1, 2).toString());
+                new JsonLocation(_rawSourceRef(true, InputStream.class), 10L, 10L, 1, 2).toString());
 
         // misc other
         Foobar srcRef = new Foobar();
         assertEquals("[Source: ("+srcRef.getClass().getName()+"); line: 1, column: 2]",
-                new JsonLocation(_rawSourceRef(srcRef), 10L, 10L, 1, 2).toString());
+                new JsonLocation(_rawSourceRef(true, srcRef), 10L, 10L, 1, 2).toString());
     }
 
     public void testTruncatedSource()
@@ -127,7 +127,7 @@
         return ContentReference.construct(true, rawSrc, -1, -1);
     }
 
-    private ContentReference _rawSourceRef(Object rawSrc) {
-        return ContentReference.rawReference(rawSrc);
+    private ContentReference _rawSourceRef(boolean textual, Object rawSrc) {
+        return ContentReference.rawReference(textual, rawSrc);
     }
 }
