diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/pom.xml ./RegMiner4APR-Regression-Bugs/BIC/pom.xml
--- ./RegMiner4APR-Regression-Bugs/WORKING/pom.xml	2026-01-02 02:18:11.603145142 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/pom.xml	2026-01-02 02:14:35.291534632 +1100
@@ -189,9 +189,6 @@
                             <goal>shade</goal>
                         </goals>
                         <configuration>
-                            <artifactSet>
-                                <excludes>commons-io:commons-io</excludes>
-                            </artifactSet>
                             <relocations>
                                 <relocation>
                                     <pattern>me.main__.util</pattern>
@@ -228,27 +225,6 @@
                             </relocations>
                         </configuration>
                     </execution>
-                    <execution>
-                        <id>apache-commons-io</id>
-                        <phase>package</phase>
-                        <goals>
-                            <goal>shade</goal>
-                        </goals>
-                        <configuration>
-                            <artifactSet>
-                                <includes>
-                                    <include>commons-io:commons-io</include>
-                                </includes>
-                            </artifactSet>
-                            <relocations>
-                                <relocation>
-                                    <pattern>org.apache.commons</pattern>
-                                    <shadedPattern>com.onarandombox.commons</shadedPattern>
-                                </relocation>
-                            </relocations>
-                            <minimizeJar>true</minimizeJar>
-                        </configuration>
-                    </execution>
                 </executions>
             </plugin>
             <plugin>
@@ -269,11 +245,6 @@
 
     <dependencies>
         <dependency>
-            <groupId>commons-io</groupId>
-            <artifactId>commons-io</artifactId>
-            <version>2.4</version>
-        </dependency>
-        <dependency>
            <groupId>org.bukkit</groupId>
            <artifactId>bukkit</artifactId>
            <version>1.13.2-R0.1-SNAPSHOT</version>
@@ -372,6 +343,12 @@
             <version>4.0.2</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>2.4</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
     <distributionManagement>
         <repository>
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/onarandombox/MultiverseCore/utils/FileUtils.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/onarandombox/MultiverseCore/utils/FileUtils.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/onarandombox/MultiverseCore/utils/FileUtils.java	2026-01-02 02:18:11.615145123 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/onarandombox/MultiverseCore/utils/FileUtils.java	2026-01-02 02:14:35.303534607 +1100
@@ -11,7 +11,11 @@
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Comparator;
 import java.util.logging.Logger;
+import java.util.stream.Stream;
 
 /**
  * File-utilities.
@@ -28,8 +32,8 @@
      * @return true if the folder was successfully deleted.
      */
     public static boolean deleteFolder(File file) {
-        try {
-            org.apache.commons.io.FileUtils.deleteDirectory(file);
+        try (Stream<Path> files = Files.walk(file.toPath())) {
+            files.sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
             return true;
         } catch (IOException e) {
             Logging.warning(e.getMessage());
@@ -44,8 +48,11 @@
      * @return true if the contents were successfully deleted
      */
     public static boolean deleteFolderContents(File file) {
-        try {
-            org.apache.commons.io.FileUtils.cleanDirectory(file);
+        try (Stream<Path> files = Files.walk(file.toPath())){
+            files.sorted(Comparator.reverseOrder())
+                    .map(Path::toFile)
+                    .filter(f -> !f.equals(file))
+                    .forEach(File::delete);
             return true;
         } catch (IOException e) {
             Logging.warning(e.getMessage());
@@ -53,8 +60,6 @@
         }
     }
 
-    private static final int COPY_BLOCK_SIZE = 1024;
-
     /**
      * Helper method to copy the world-folder.
      * @param source Source-File
@@ -64,8 +69,16 @@
      * @return if it had success
      */
     public static boolean copyFolder(File source, File target, Logger log) {
-        try {
-            org.apache.commons.io.FileUtils.copyDirectory(source, target);
+        Path sourcePath = source.toPath();
+        Path destPath = target.toPath();
+        try (Stream<Path> files = Files.walk(source.toPath())) {
+            files.forEachOrdered(src -> {
+                try {
+                    Files.copy(src, sourcePath.resolve(destPath.relativize(src)));
+                } catch (IOException e) {
+                    log.warning(e.getMessage());
+                }
+            });
             return true;
         } catch (IOException e) {
             log.warning(e.getMessage());
