diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/cactoos/func/RepeatedFunc.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/cactoos/func/RepeatedFunc.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/cactoos/func/RepeatedFunc.java	2025-12-22 04:05:32.000000000 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/cactoos/func/RepeatedFunc.java	2025-12-22 04:05:16.000000000 +1100
@@ -24,6 +24,7 @@
 package org.cactoos.func;
 
 import org.cactoos.Func;
+import org.cactoos.text.FormattedText;
 
 /**
  * Func that repeats its calculation a few times before
@@ -67,6 +68,14 @@
         for (int idx = 0; idx < this.times; ++idx) {
             result = this.func.apply(input);
         }
+        if (result == null) {
+            throw new IllegalArgumentException(
+                new FormattedText(
+                    "Repeat counter is equal or less than zero: %d",
+                    this.times
+                ).asString()
+            );
+        }
         return result;
     }
 
