diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/adobe/epubcheck/ctc/css/EpubCSSCheckCSSHandler.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/adobe/epubcheck/ctc/css/EpubCSSCheckCSSHandler.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/adobe/epubcheck/ctc/css/EpubCSSCheckCSSHandler.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/adobe/epubcheck/ctc/css/EpubCSSCheckCSSHandler.java
@@ -559,11 +567,17 @@
           switch (quantity.getUnit())
           {
             case EMS:
+            case EXS:
             case REMS:
             case PERCENTAGE:
               break;
+            case LENGTH:
+              getReport().message(id, getCorrectedEPUBLocation(path, declaration.getLocation().getLine(), declaration.getLocation().getColumn(), declaration.toCssString()), construct.toCssString());
+              break;
             default:
-              getReport().message(id, getCorrectedEPUBLocation(path, declaration.getLocation().getLine(), declaration.getLocation().getColumn(), declaration.toCssString()));
+              getReport().message(MessageId.CSS_020, getCorrectedEPUBLocation(path, declaration.getLocation().getLine(), declaration.getLocation().getColumn(), declaration.toCssString()), construct.toCssString());
               break;
           }
         }
diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/adobe/epubcheck/messages/MessageDictionary.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/adobe/epubcheck/messages/MessageDictionary.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/adobe/epubcheck/messages/MessageDictionary.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/adobe/epubcheck/messages/MessageDictionary.java
@@ -112,7 +112,7 @@
       map.put(MessageId.CSS_016, Severity.SUPPRESSED);
       map.put(MessageId.CSS_017, Severity.WARNING);
       map.put(MessageId.CSS_019, Severity.WARNING);
-      map.put(MessageId.CSS_020, Severity.USAGE);
+      map.put(MessageId.CSS_020, Severity.ERROR);
       map.put(MessageId.CSS_021, Severity.USAGE);
       map.put(MessageId.CSS_022, Severity.USAGE);
       map.put(MessageId.CSS_023, Severity.USAGE);