diff -r -U 3 RegressionBug-147/WORKING/src/main/java/mikera/matrixx/Matrix.java RegressionBug-147/BIC/src/main/java/mikera/matrixx/Matrix.java
--- RegressionBug-147/WORKING/src/main/java/mikera/matrixx/Matrix.java	2026-04-03 02:09:55.578719399 +1100
+++ RegressionBug-147/BIC/src/main/java/mikera/matrixx/Matrix.java	2026-04-03 02:08:00.118659899 +1100
@@ -522,6 +526,16 @@
 	}
 
 	@Override
+	public void set(AVector a) {
+		if ((rowCount() != a.length())) { throw new IllegalArgumentException(
+				ErrorMessages.incompatibleBroadcast(a, this)); }
+		a.getElements(data, 0);
+		for (int i = 1; i < rows; i++) {
+			System.arraycopy(data, 0, data, i * cols, cols);
+		}
+	}
 
 	@Override
 	public void getElements(double[] dest, int offset) {

