diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/logicng/formulas/FormulaFactory.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/logicng/formulas/FormulaFactory.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/logicng/formulas/FormulaFactory.java	2024-09-08 01:36:07.425761748 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/logicng/formulas/FormulaFactory.java	2024-09-08 01:36:07.361761299 +1000
@@ -276,6 +278,8 @@
       return this.not(left);
     if (left.equals(right))
       return this.verum();
+    if (left.negate().equals(right))
+      return left.negate();
     final Pair<Formula, Formula> key = new Pair<>(left, right);
     Implication implication = this.implications.get(key);
     if (implication == null) {