diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/input/BomInput.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/input/BomInput.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/input/BomInput.java	2024-09-08 01:36:15.281816766 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/input/BomInput.java	2024-09-08 01:36:15.265816654 +1000
@@ -117,25 +120,23 @@
 
 	@Override
 	public final int read() throws IOException {
-		if (bytesRead > 0) {
-			int out = b1;
-			if (bytesRead == 2) {
-				out = b2;
-			} else if (bytesRead == 3) {
-				out = b3;
-			} else if (bytesRead == 4) {
-				out = b4;
-			}
+		if (bytesRead > 0 && bytesRead > consumed) {
+			int out = bytes[consumed];
 
 			// Ensures that if the original input stream returned a byte, it will be consumed.
 			// In case of exceptions, bytes produced prior to the exception will still be returned.
 			// Once the last byte has been consumed, the original exception will be thrown.
-			if (--bytesRead == 0 && exception != null) {
+			if (++consumed == bytesRead && exception != null) {
 				throw exception;
 			}
 			return out;
 		}
-		return input.read();
+		if (consumed == bytesRead) {
+			consumed++;
+			return -1;
+		}

+		throw new BytesProcessedNotification(input, encoding);
 	}
 
 	/**
