diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/ClassBuilderInfo.java ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/ClassBuilderInfo.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/ClassBuilderInfo.java	2025-12-19 20:47:34.105845747 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/ClassBuilderInfo.java	2025-12-19 20:39:08.698133619 +1100
@@ -19,12 +19,12 @@
 
 /**
  * Helper to create Java class/interface files with proper formatting.
- * 
+ *
  * @author hypfvieh
  * @since v3.0.1 - 2018-12-22
  */
 public class ClassBuilderInfo {
-    
+
     /** Set of reserved words in Java. */
     private static final Set<String> RESERVED = new HashSet<>();
     static {
@@ -80,7 +80,7 @@
         RESERVED.add("volatile");
         RESERVED.add("while");
     }
-    
+
     /** Imported files for this class. */
     private final Set<String>            imports               = new TreeSet<>();
     /** Members/Fields of this class. */
@@ -98,7 +98,7 @@
     private String                       className;
     /** Package of this class. */
     private String                       packageName;
-    
+
     /** Package name used by DBus. */
     private String                       dbusPackageName;
 
@@ -170,7 +170,7 @@
     public List<ClassConstructor> getConstructors() {
         return constructors;
     }
-    
+
     /**
      * Create the Java source for the class information provided.
      * @return String
@@ -182,7 +182,7 @@
 
     /**
      * Create the Java source for the class information provided.
-     * 
+     *
      * @param _staticClass this is static inner class
      * @param _otherImports this class needs additional imports (e.g. due to inner class)
      * @return
@@ -198,7 +198,7 @@
         if (_otherImports != null) {
             allImports.addAll(_otherImports);
         }
-        
+
         if (!_staticClass) {
             content.add("package " + getPackageName() + ";");
             content.add("");
@@ -214,7 +214,7 @@
             allImports.add(DBusInterfaceName.class.getName());
             content.add(classIndent + "@" + DBusInterfaceName.class.getSimpleName() + "(\"" + getDbusPackageName() + "\")");
         }
-        
+
         String bgn = classIndent + "public " + (_staticClass ? "static " : "") + (getClassType() == ClassType.INTERFACE ? "interface" : "class");
         bgn += " " + getClassName();
         if (getExtendClass() != null) {
@@ -249,14 +249,14 @@
             }
             content.add(memberIndent + "private " + (member.isFinalMember() ? "final " : "") + memberType + " "
                     + member.getName() + ";");
-        
+
         }
 
         if (!getConstructors().isEmpty()) {
             content.add("");
             for (ClassConstructor constructor : getConstructors()) {
                 String outerIndent = _staticClass ? "        " : "    ";
-                String cstr = outerIndent + getClassName() + "(";
+                String cstr = outerIndent + "public " + getClassName() + "(";
                 if (!constructor.getSuperArguments().isEmpty()) {
                     cstr += constructor.getSuperArguments().entrySet().stream().map(e -> e.getValue() + " " + e.getKey()).collect(Collectors.joining(", "));
                     if (!constructor.getArguments().isEmpty()) {
@@ -264,19 +264,20 @@
                     }
                 }
                 if (!constructor.getArguments().isEmpty()) {
-                    cstr += constructor.getArguments().entrySet().stream().map(e -> e.getValue() + " " + e.getKey()).collect(Collectors.joining(", "));
+                    cstr += constructor.getArguments().entrySet().stream()
+                            .map(e -> TypeConverter.getProperJavaClass(e.getValue(), allImports) + " " + e.getKey()).collect(Collectors.joining(", "));
                 }
-                
+
                 if (constructor.getThrowArguments().isEmpty()) {
                     cstr += ") {";
                 } else {
                     cstr += ") throws " + String.join(", ", constructor.getThrowArguments()) + " {";
                 }
-                
+
                 content.add(cstr);
-                
+
                 String innerIndent = _staticClass ? "            " : "        ";
-                
+
                 if (!constructor.getSuperArguments().isEmpty()) {
                     content.add(innerIndent + "super(" + String.join(", ", constructor.getSuperArguments().keySet()) + ");");
                 }
@@ -316,11 +317,11 @@
         content.add("");
 
         for (ClassMethod mth : getMethods()) {
-        	
+
         	if (!mth.getAnnotations().isEmpty()) {
         		content.addAll(mth.getAnnotations().stream().map(a -> memberIndent + a).collect(Collectors.toList()));
         	}
-        	
+
             String clzMth = memberIndent + "public " + (mth.getReturnType() == null ? "void " : TypeConverter.getProperJavaClass(mth.getReturnType(), allImports) + " ");
             clzMth += mth.getName() + "(";
             if (!mth.getArguments().isEmpty()) {
@@ -336,7 +337,7 @@
             content.addAll(inner.createClassFileContent(true, allImports));
             allImports.addAll(inner.getImports()); // collect additional imports which may have been added by inner class
         }
-        
+
         content.add(classIndent + "}");
 
         // write imports to resulting content if this is not a inner class
@@ -359,7 +360,7 @@
         if (getPackageName() == null) { // no package name
         	return getClassName() + ".java";
         }
-        
+
         return getPackageName().replace(".", File.separator) + File.separator + getClassName() + ".java";
     }
 
@@ -390,16 +391,16 @@
 
     /**
      * Contains information about annotation to place on classes, members or methods.
-     * 
+     *
      * @author hypfvieh
      * @since v3.2.1 - 2019-11-13
      */
     public static class AnnotationInfo {
         /** Annotation class. */
         private final Class<? extends Annotation> annotationClass;
-        /** Annotation params (e.g. value = "foo", key = "bar"). */ 
+        /** Annotation params (e.g. value = "foo", key = "bar"). */
         private final String annotationParams;
-        
+
         public AnnotationInfo(Class<? extends Annotation> _annotationClass, String _annotationParams) {
             annotationClass = _annotationClass;
             annotationParams = _annotationParams;
@@ -413,10 +414,10 @@
             return annotationParams;
         }
     }
-    
+
     /**
      * Pojo which represents a class method.
-     * 
+     *
      * @author hypfvieh
      * @since v3.0.1 - 2018-12-20
      */
@@ -431,7 +432,7 @@
         private final Map<String, String> arguments = new LinkedHashMap<>();
         /** List of annotations for this method. */
         private final List<String> annotations = new ArrayList<>();
-        
+
         public ClassMethod(String _name, String _returnType, boolean _finalMethod) {
             name = _name;
             returnType = _returnType;
@@ -457,12 +458,12 @@
 		public List<String> getAnnotations() {
 			return annotations;
 		}
-        
+
     }
 
     /**
      * Pojo which represents a class member/field.
-     * 
+     *
      * @author hypfvieh
      * @since v3.0.1 - 2018-12-20
      */
@@ -480,7 +481,7 @@
 
         public ClassMember(String _name, String _type, boolean _finalMember) {
             // repair reserved words by adding 'Param' as appendix
-        	name = RESERVED.contains(_name) ? _name + "Param" : _name;
+        	name = RESERVED.contains(_name) ? _name + "param" : _name;
             type = _type;
             finalMember = _finalMember;
         }
@@ -509,27 +510,27 @@
 
     /**
      * Pojo which represents a class constructor.
-     * 
+     *
      * @author hypfvieh
      * @since v3.0.1 - 2018-12-20
      */
     public static class ClassConstructor {
-    	/** List of arguments for the constructor. Key is argument name, value is argument type. */
+    	/** Map of arguments for the constructor. Key is argument name, value is argument type. */
         private final Map<String, String> arguments = new LinkedHashMap<>();
-        /** List of arguments for the super-constructor. Key is argument name, value is argument type. */
+        /** Map of arguments for the super-constructor. Key is argument name, value is argument type. */
         private final Map<String, String> superArguments = new LinkedHashMap<>();
-        
+
         /** List of throws arguments. */
         private final List<String> throwArguments = new ArrayList<>();
-        
+
         public List<String> getThrowArguments() {
             return throwArguments;
         }
-        
+
         public Map<String, String> getArguments() {
             return arguments;
         }
-        
+
         public Map<String, String> getSuperArguments() {
             return superArguments;
         }
@@ -537,7 +538,7 @@
 
     /**
      * Enum to define either the {@link ClassBuilderInfo} is for a CLASS or an INTERFACE.
-     * 
+     *
      * @author hypfvieh
      * @since v3.0.1 - 2018-12-20
      */
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGenerator.java ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGenerator.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGenerator.java	2025-12-19 20:47:34.105845747 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/main/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGenerator.java	2025-12-19 20:39:08.698133619 +1100
@@ -95,13 +95,19 @@
 
         Map<File, String> filesAndContents = new LinkedHashMap<>();
 
+        boolean noBusnameGiven = StringUtil.isBlank(busName);
 
         for (Element ife : interfaceElements) {
-            if (!StringUtil.isBlank(busName) && ife.getAttribute("name").startsWith(busName)) {
+            String nameAttrib = ife.getAttribute("name");
+            if (!noBusnameGiven && !nameAttrib.startsWith(busName)) { // busname was set, and current element does not match -> skip
+                logger.info("Skipping: {} - does not match given busName: {}", nameAttrib, busName);
+                continue;
+            } else if (noBusnameGiven) { // no busname given, take all
+                // take all interfaces
                 filesAndContents.putAll(extractAll(ife));
                 continue;
             }
-            // take all interfaces
+            // busname given and matching
             filesAndContents.putAll(extractAll(ife));
         }
 
@@ -130,7 +136,7 @@
      * Extract all methods/signals etc. from the given interface element.
      *
      * @param _ife interface element
-     * @return List of files and their contents
+     * @return Map of files and their contents
      * @throws IOException when reading xml fails
      * @throws DBusException when DBus fails
      */
@@ -233,15 +239,15 @@
         for (Entry<String, String> e : args.entrySet()) {
             argsMap.put("_" + e.getKey(), e.getValue());
         }
-        
+
         classConstructor.getArguments().putAll(argsMap);
         classConstructor.getThrowArguments().add(DBusException.class.getSimpleName());
-        
+
         classConstructor.getSuperArguments().put("_path", "String");
         classConstructor.getSuperArguments().put("_interfaceName", "String");
 
         innerClass.getConstructors().add(classConstructor);
-        
+
 
         return additionalClasses;
     }
@@ -445,7 +451,7 @@
         if (objectPath == null) {
             objectPath = "/";
         }
-        
+
         if (outputDir == null) {
             throw new RuntimeException("No output directory (--outputDir) given!");
         }
@@ -454,47 +460,45 @@
 
         String introspectionData = null;
 
-        if (!StringUtil.isBlank(busName)) {
-            if (!StringUtil.isBlank(inputFile)) {
-                File file = new File(inputFile);
-                if (!file.exists()) {
-                    logger.error("Given input file {} does not exist", file);
-                    System.exit(1);
-                }
-                introspectionData = FileIoUtil.readFileToString(file);
-            } else {
-                try {
-                    logger.info("Introspecting: { Interface: {}, Busname: {} }", objectPath, busName);
-    
-                    DBusConnection conn = DBusConnection.getConnection(busType);
-    
-                    Introspectable in = conn.getRemoteObject(busName, objectPath, Introspectable.class);
-                    introspectionData = in.Introspect();
-                    if (StringUtil.isBlank(introspectionData)) {
-                        logger.error("Failed to get introspection data");
-                        System.exit(1);
-                    }
-                    conn.disconnect();
-                } catch (DBusExecutionException | DBusException _ex) {
-                    logger.error("Failure in DBus Communications. ", _ex);
-                    System.exit(1);
-    
-                }
+        if (!StringUtil.isBlank(inputFile)) {
+            File file = new File(inputFile);
+            if (!file.exists()) {
+                logger.error("Given input file {} does not exist", file);
+                System.exit(1);
             }
-            
-            InterfaceCodeGenerator ci2 = new InterfaceCodeGenerator(introspectionData, objectPath, busName);
+            introspectionData = FileIoUtil.readFileToString(file);
+        } else if (!StringUtil.isBlank(busName)) {
             try {
+                logger.info("Introspecting: { Interface: {}, Busname: {} }", objectPath, busName);
+
+                DBusConnection conn = DBusConnection.getConnection(busType);
+
+                Introspectable in = conn.getRemoteObject(busName, objectPath, Introspectable.class);
+                introspectionData = in.Introspect();
+                if (StringUtil.isBlank(introspectionData)) {
+                    logger.error("Failed to get introspection data");
+                    System.exit(1);
+                }
+                conn.disconnect();
+            } catch (DBusExecutionException | DBusException _ex) {
+                logger.error("Failure in DBus Communications. ", _ex);
+                System.exit(1);
 
-                Map<File, String> analyze = ci2.analyze(ignoreDtd);
-                writeToFile(outputDir, analyze);
-                logger.info("Interface creation finished");
-            } catch (Exception _ex) {
-                logger.error("Error while analyzing introspection data", _ex);
             }
         } else {
             logger.error("Busname missing!");
             System.exit(1);
         }
+
+        InterfaceCodeGenerator ci2 = new InterfaceCodeGenerator(introspectionData, objectPath, busName);
+        try {
+
+            Map<File, String> analyze = ci2.analyze(ignoreDtd);
+            writeToFile(outputDir, analyze);
+            logger.info("Interface creation finished");
+        } catch (Exception _ex) {
+            logger.error("Error while analyzing introspection data", _ex);
+        }
     }
 
     private static void version() {
@@ -513,6 +517,9 @@
         System.out.println("        --enable-dtd-validation          Enable DTD validation of introspection XML");
         System.out.println("        --version                        Show version information");
         System.out.println("        --help                           Show this help");
+        System.out.println("");
+        System.out.println("If --inputFile is given busname object argument can be skipped, that will force the util to extract all interfaces found in the given file.");
+        System.out.println("If a busname (not empty or blank) is given, then only interfaces starting with the given busname will be extracted.");
     }
 
     static enum DbusInterfaceToFqcn {
@@ -532,10 +539,10 @@
 
             map.put(DbusInterfaceToFqcn.CLASSNAME, StringUtil.upperCaseFirstChar(className));
             map.put(DbusInterfaceToFqcn.PACKAGENAME, packageName.toLowerCase());
-            
+
             if (!packageName.equals(packageName.toLowerCase())) {
-                map.put(DbusInterfaceToFqcn.ORIG_PKGNAME, packageName);    
-                map.put(DbusInterfaceToFqcn.DBUS_INTERFACE_NAME, packageName + "." + className); 
+                map.put(DbusInterfaceToFqcn.ORIG_PKGNAME, packageName);
+                map.put(DbusInterfaceToFqcn.DBUS_INTERFACE_NAME, packageName + "." + className);
             }
             return map;
         }
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/test/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGeneratorTest.java ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/test/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGeneratorTest.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/dbus-java-utils/src/test/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGeneratorTest.java	2025-12-19 20:47:34.105845747 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/dbus-java-utils/src/test/java/org/freedesktop/dbus/utils/generator/InterfaceCodeGeneratorTest.java	2025-12-19 20:39:08.702133617 +1100
@@ -25,6 +25,7 @@
 
         Logger logger = LoggerFactory.getLogger(InterfaceCodeGenerator.class);
 
+
         if (!StringUtils.isBlank(busName)) {
             String introspectionData = FileIoUtil.readFileToString("src/test/resources/CreateInterface/firewall/org.fedoraproject.FirewallD1.xml");
 
