diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/AbstractParser.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/AbstractParser.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/AbstractParser.java	2025-12-27 06:01:03.565797975 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/AbstractParser.java	2025-12-27 05:59:25.485857591 +1100
@@ -102,7 +103,8 @@
 	public final void parse(Reader reader) {
 		beginParsing(reader);
 		try {
-			while (!context.stopped && (ch = input.nextChar()) != '\0') {
+			while (!context.stopped) {
+				ch = input.nextChar();
 				if (ch == comment) {
 					input.skipLines(1);
 					continue;
@@ -119,6 +121,24 @@
 					}
 				}
 			}
+		} catch (EOFException ex) {
+			if (output.column != 0) {
+				if (output.appender.length() > 0) {
+					output.valueParsed();
+				} else {
+					output.emptyParsed();
+				}
+				String[] row = output.rowParsed();
+				if (row != null) {
+					processor.rowProcessed(row, context);
+				}
+			} else if (output.appender.length() > 0) {
+				output.valueParsed();
+				String[] row = output.rowParsed();
+				if (row != null) {
+					processor.rowProcessed(row, context);
+				}
+			}
 		} catch (Exception ex) {
 			throw new TextParsingException(context, ex);
 		} finally {
@@ -147,7 +167,8 @@
 	 */
 	public final String[] parseNext() {
 		try {
-			while (!context.stopped && (ch = input.nextChar()) != '\0') {
+			while (!context.stopped) {
+				ch = input.nextChar();
 				if (ch == comment) {
 					input.skipLines(1);
 					continue;
@@ -166,6 +187,13 @@
 			}
 			stopParsing();
 			return null;
+		} catch (EOFException ex) {
+			if (output.appender.length() > 0) {
+				output.valueParsed();
+			}
+			String[] row = output.rowParsed();
+			stopParsing();
+			return row;
 		} catch (Exception ex) {
 			try {
 				throw new TextParsingException(context, ex);
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/input/AbstractCharInputReader.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/input/AbstractCharInputReader.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/input/AbstractCharInputReader.java	2025-12-27 06:01:03.569797973 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/input/AbstractCharInputReader.java	2025-12-27 05:59:25.489857589 +1100
@@ -116,7 +116,7 @@
 	@Override
 	public final char nextChar() {
 		if (length == -1) {
-			return '\0';
+			throw new EOFException();
 		}
 
 		char ch = buffer[i - 1];
@@ -125,7 +125,7 @@
 			if (length != -1) {
 				updateBuffer();
 			} else {
-				return '\0';
+				throw new EOFException();
 			}
 		}
 
@@ -140,7 +140,7 @@
 					if (length != -1) {
 						updateBuffer();
 					} else {
-						return '\0';
+						throw new EOFException();
 					}
 				}
 
@@ -172,10 +172,14 @@
 		int expectedLineCount = this.lineCount + lines;
 
 		char ch = '\0';
-		do {
-			ch = nextChar();
-		} while (lineCount < expectedLineCount && ch != '\0');
-		if (ch == '\0' && lineCount < lines) {
+		try {
+			do {
+				ch = nextChar();
+			} while (lineCount < expectedLineCount && ch != '\0');
+			if (lineCount < lines) {
+				throw new IllegalArgumentException("Unable to skip " + lines + " lines from line " + (expectedLineCount - lines) + ". End of input reached");
+			}
+		} catch (EOFException ex) {
 			throw new IllegalArgumentException("Unable to skip " + lines + " lines from line " + (expectedLineCount - lines) + ". End of input reached");
 		}
 	}
Only in ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/input: EOFException.java
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/ParserOutput.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/ParserOutput.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/common/ParserOutput.java	2025-12-27 06:01:03.565797975 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/common/ParserOutput.java	2025-12-27 05:59:25.485857591 +1100
@@ -45,7 +45,7 @@
 	 * Calls to {@link ParserOutput#valueParsed} and  {@link ParserOutput#emptyParsed} will increase the column count.
 	 * Calls to {@link ParserOutput#clear} will reset it to zero.
 	 */
-	private int column = 0;
+	int column = 0;
 
 	/**
 	 * Stores the values parsed for a record.
@@ -119,6 +119,7 @@
 				//skips the header row. We want to use the headers defined in the settings.
 				if (settings.isHeaderExtractionEnabled()) {
 					Arrays.fill(parsedValues, null);
+					column = 0;
 					return null;
 				}
 			}
@@ -134,10 +135,12 @@
 						reorderedValues[i] = parsedValues[index];
 					}
 				}
+				column = 0;
 				return reorderedValues;
 			} else {
 				String[] out = new String[column];
 				System.arraycopy(parsedValues, 0, out, 0, column);
+				column = 0;
 				return out;
 			}
 		} else if (!skipEmptyLines) { //no values were parsed, but we are not skipping empty lines
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/csv/CsvParser.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/csv/CsvParser.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/csv/CsvParser.java	2025-12-27 06:01:03.573797970 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/csv/CsvParser.java	2025-12-27 05:59:25.493857586 +1100
@@ -63,7 +63,7 @@
 			skipWhitespace();
 		}
 
-		while (ch != newLine && ch != '\0') {
+		while (ch != newLine) {
 			parseField();
 			if (ch != newLine) {
 				ch = input.nextChar();
@@ -76,12 +76,12 @@
 
 	private void parseValue() {
 		if (ignoreTrailingWhitespace) {
-			while (ch != delimiter && ch != newLine && ch != '\0') {
+			while (ch != delimiter && ch != newLine) {
 				output.appender.appendIgnoringWhitespace(ch);
 				ch = input.nextChar();
 			}
 		} else {
-			while (ch != delimiter && ch != newLine && ch != '\0') {
+			while (ch != delimiter && ch != newLine) {
 				output.appender.append(ch);
 				ch = input.nextChar();
 			}
@@ -92,7 +92,7 @@
 		char prev = '\0';
 		ch = input.nextChar();
 
-		while (!(prev == quote && (ch == delimiter || ch <= ' ')) && ch != '\0') {
+		while (!(prev == quote && (ch == delimiter || ch <= ' '))) {
 			if (ch != quote) {
 				output.appender.append(ch);
 				prev = ch;
@@ -129,7 +129,7 @@
 	}
 
 	private void skipWhitespace() {
-		while (ch <= ' ' && ch != delimiter && ch != newLine && ch != '\0') {
+		while (ch <= ' ' && ch != delimiter && ch != newLine) {
 			ch = input.nextChar();
 		}
 	}
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/fixed/FixedWidthParser.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/fixed/FixedWidthParser.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/fixed/FixedWidthParser.java	2025-12-27 06:01:03.573797970 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/fixed/FixedWidthParser.java	2025-12-27 05:59:25.493857586 +1100
@@ -100,31 +100,31 @@
 	}
 
 	private void skipToNewLine() {
-		while (ch != newLine && ch != '\0') {
+		while (ch != newLine) {
 			ch = input.nextChar();
 		}
 	}
 
 	private void skipPadding() {
-		while (ch == padding && length-- > 0 && ch != '\0') {
+		while (ch == padding && length-- > 0) {
 			ch = input.nextChar();
 		}
 	}
 
 	private void skipWhitespace() {
-		while (ch <= ' ' && length-- > 0 && ch != '\0') {
+		while (ch <= ' ' && length-- > 0) {
 			ch = input.nextChar();
 		}
 	}
 
 	private void readValueUntilNewLine() {
 		if (ignoreTrailingWhitespace) {
-			while (length-- > 0 && ch != newLine && ch != '\0') {
+			while (length-- > 0 && ch != newLine) {
 				output.appender.appendIgnoringWhitespaceAndPadding(ch);
 				ch = input.nextChar();
 			}
 		} else {
-			while (length-- > 0 && ch != newLine && ch != '\0') {
+			while (length-- > 0 && ch != newLine) {
 				output.appender.appendIgnoringPadding(ch);
 				ch = input.nextChar();
 			}
@@ -133,12 +133,12 @@
 
 	private void readValue() {
 		if (ignoreTrailingWhitespace) {
-			while (length-- > 0 && ch != '\0') {
+			while (length-- > 0) {
 				output.appender.appendIgnoringWhitespaceAndPadding(ch);
 				ch = input.nextChar();
 			}
 		} else {
-			while (length-- > 0 && ch != '\0') {
+			while (length-- > 0) {
 				output.appender.appendIgnoringPadding(ch);
 				ch = input.nextChar();
 			}
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/tsv/TsvParser.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/tsv/TsvParser.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/tsv/TsvParser.java	2025-12-27 06:01:03.573797970 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/tsv/TsvParser.java	2025-12-27 05:59:25.497857584 +1100
@@ -59,7 +59,7 @@
 			skipWhitespace();
 		}
 
-		while (ch != newLine && ch != '\0') {
+		while (ch != newLine) {
 			parseField();
 			if (ch != newLine) {
 				ch = input.nextChar();
@@ -79,7 +79,7 @@
 			output.emptyParsed();
 		} else {
 			if (ignoreTrailingWhitespace) {
-				while (ch != '\t' && ch != newLine && ch != '\0') {
+				while (ch != '\t' && ch != newLine) {
 					if (ch == escapeChar) {
 						ch = input.nextChar();
 						if (ch == 't') {
@@ -92,7 +92,7 @@
 							output.appender.appendIgnoringWhitespace('\r');
 						} else {
 							output.appender.append(escapeChar);
-							if (ch == newLine || ch == '\t' || ch == '\0') {
+							if (ch == newLine || ch == '\t') {
 								break;
 							}
 							output.appender.appendIgnoringWhitespace(ch);
@@ -104,7 +104,7 @@
 					}
 				}
 			} else {
-				while (ch != '\t' && ch != newLine && ch != '\0') {
+				while (ch != '\t' && ch != newLine) {
 					if (ch == escapeChar) {
 						ch = input.nextChar();
 						if (ch == 't') {
@@ -117,7 +117,7 @@
 							output.appender.append('\r');
 						} else {
 							output.appender.append(escapeChar);
-							if (ch == newLine || ch == '\0' || ch == '\t') {
+							if (ch == newLine || ch == '\t') {
 								break;
 							}
 							output.appender.append(ch);
@@ -134,7 +134,7 @@
 	}
 
 	private void skipWhitespace() {
-		while (ch != '\t' && ch <= ' ' && ch != newLine && ch != '\0') {
+		while (ch != '\t' && ch <= ' ' && ch != newLine) {
 			ch = input.nextChar();
 		}
 	}
