diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/csv/CsvFormatDetector.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/csv/CsvFormatDetector.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/com/univocity/parsers/csv/CsvFormatDetector.java	2024-09-08 01:36:14.481811163 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/com/univocity/parsers/csv/CsvFormatDetector.java	2024-09-08 01:36:14.457810995 +1000
@@ -325,7 +325,7 @@
 	}
 
 	private boolean isSymbol(char ch) {
-		return ch != comment && !Character.isLetterOrDigit(ch) && (ch == '\t' || ch > ' ');
+		return ch != comment && !Character.isLetterOrDigit(ch) && (ch == '\t' || ch >= ' ');
 	}
 
 	/**