diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/cactoos/io/WriterAsOutputStream.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/cactoos/io/WriterAsOutputStream.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/cactoos/io/WriterAsOutputStream.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/cactoos/io/WriterAsOutputStream.java
@@ -171,6 +171,11 @@
         }
     }
 
+    @Override
+    public void close() throws IOException {
+        this.writer.close();
+    }

     /**
      * Write a portion from the buffer.
      * @param buffer The buffer