diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/jsoup/parser/CharacterReader.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/jsoup/parser/CharacterReader.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/jsoup/parser/CharacterReader.java	2025-12-14 15:54:20.000000000 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/jsoup/parser/CharacterReader.java	2025-12-14 15:53:57.000000000 +1100
@@ -48,14 +48,17 @@
             return;
 
         try {
-            readerPos += bufPos;
             reader.skip(bufPos);
             reader.mark(maxBufferLen);
-            bufLength = reader.read(charBuf);
+            final int read = reader.read(charBuf);
             reader.reset();
-            bufPos = 0;
-            bufMark = 0;
-            bufSplitPoint = bufLength > readAheadLimit ? readAheadLimit : bufLength;
+            if (read != -1) {
+                bufLength = read;
+                readerPos += bufPos;
+                bufPos = 0;
+                bufMark = 0;
+                bufSplitPoint = bufLength > readAheadLimit ? readAheadLimit : bufLength;
+            }
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
@@ -74,6 +77,11 @@
      * @return true if nothing left to read.
      */
     public boolean isEmpty() {
+        bufferUp();
+        return bufPos >= bufLength;
+    }
+
+    private boolean isEmptyNoBufferUp() {
         return bufPos >= bufLength;
     }
 
@@ -83,12 +91,12 @@
      */
     public char current() {
         bufferUp();
-        return isEmpty() ? EOF : charBuf[bufPos];
+        return isEmptyNoBufferUp() ? EOF : charBuf[bufPos];
     }
 
     char consume() {
         bufferUp();
-        char val = isEmpty() ? EOF : charBuf[bufPos];
+        char val = isEmptyNoBufferUp() ? EOF : charBuf[bufPos];
         bufPos++;
         return val;
     }
@@ -281,7 +289,7 @@
             else
                 break;
         }
-        while (!isEmpty()) {
+        while (!isEmptyNoBufferUp()) {
             char c = charBuf[bufPos];
             if (c >= '0' && c <= '9')
                 bufPos++;
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/org/jsoup/parser/CharacterReaderTest.java ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/org/jsoup/parser/CharacterReaderTest.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/org/jsoup/parser/CharacterReaderTest.java	2025-12-14 15:54:20.000000000 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/org/jsoup/parser/CharacterReaderTest.java	2025-12-14 15:53:57.000000000 +1100
@@ -1,7 +1,10 @@
 package org.jsoup.parser;
 
+import org.junit.Ignore;
 import org.junit.Test;
 
+import java.io.StringReader;
+
 import static org.junit.Assert.*;
 
 /**
@@ -268,5 +271,18 @@
         assertTrue(r.isEmpty());
     }
 
+    @Ignore
+    @Test
+    public void notEmptyAtBufferSplitPoint() {
+        CharacterReader r = new CharacterReader(new StringReader("How about now"), 3);
+        assertEquals("How", r.consumeTo(' '));
+        assertFalse("Should not be empty", r.isEmpty());
+
+        assertEquals(' ', r.consume());
+        assertFalse(r.isEmpty());
+
+        // todo - current consume to won't expand buffer. impl buffer extension and test
+    }
+
 
 }
