diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/org/traccar/protocol/EelinkProtocolDecoder.java ./RegMiner4APR-Regression-Bugs/BIC/src/org/traccar/protocol/EelinkProtocolDecoder.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/org/traccar/protocol/EelinkProtocolDecoder.java	2026-02-19 15:06:44.023366691 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/org/traccar/protocol/EelinkProtocolDecoder.java	2026-02-19 16:06:11.919507802 +1100
@@ -131,22 +131,37 @@
 
         position.setValid((buf.readUnsignedByte() & 0x01) != 0);
 
-        if (type == MSG_ALARM) {
-            position.set(Position.KEY_ALARM, decodeAlarm(buf.readUnsignedByte()));
-        }
+        if (type == MSG_GPS) {
 
-        if (buf.readableBytes() >= 2) {
-            decodeStatus(position, buf.readUnsignedShort());
-        }
+            if (buf.readableBytes() >= 2) {
+                decodeStatus(position, buf.readUnsignedShort());
+            }

+            if (buf.readableBytes() >= 2 * 4) {
+                position.set(Position.KEY_BATTERY, buf.readUnsignedShort() * 0.001);
+                position.set(Position.KEY_RSSI, buf.readUnsignedShort());
+                position.set(Position.PREFIX_ADC + 1, buf.readUnsignedShort());
+                position.set(Position.PREFIX_ADC + 2, buf.readUnsignedShort());
+            }
+
+        } else if (type == MSG_ALARM) {
+
+            position.set(Position.KEY_ALARM, decodeAlarm(buf.readUnsignedByte()));
 
-        if (buf.readableBytes() >= 2 * 4) {
+        } else if (type == MSG_STATE) {
 
-            position.set(Position.KEY_BATTERY, buf.readUnsignedShort() * 0.001);
+            int statusType = buf.readUnsignedByte();
 
-            position.set(Position.KEY_RSSI, buf.readUnsignedShort());
+            position.set(Position.KEY_EVENT, statusType);
 
-            position.set(Position.PREFIX_ADC + 1, buf.readUnsignedShort());
-            position.set(Position.PREFIX_ADC + 2, buf.readUnsignedShort());
+            if (statusType == 0x01 || statusType == 0x02 || statusType == 0x03) {
+                buf.readUnsignedInt(); // device time
+                decodeStatus(position, buf.readUnsignedShort());
+            }
 
         }
 
