diff -r -U 3 RegressionBug-100/WORKING/src/main/java/inra/ijpb/measure/region2d/AverageThickness.java RegressionBug-100/BIC/src/main/java/inra/ijpb/measure/region2d/AverageThickness.java
--- RegressionBug-100/WORKING/src/main/java/inra/ijpb/measure/region2d/AverageThickness.java
+++ RegressionBug-100/BIC/src/main/java/inra/ijpb/measure/region2d/AverageThickness.java
@@ -111,12 +111,19 @@
         {
             for (int x = 0; x < sizeX; x++)
             {
                 int label = (int) skeleton.getf(x, y);

                 if (label == 0)
                 {
                     continue;
                 }

+                if (labelIndices.containsKey(label))
+                {
+                    continue;
+                }
 
                 int index = labelIndices.get(label);
                 
