diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/io/github/classgraph/ClasspathElementZip.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/io/github/classgraph/ClasspathElementZip.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/io/github/classgraph/ClasspathElementZip.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/io/github/classgraph/ClasspathElementZip.java
@@ -446,7 +446,8 @@
             if (parentMatchStatus == ScanSpecPathMatch.HAS_WHITELISTED_PATH_PREFIX
                     || parentMatchStatus == ScanSpecPathMatch.AT_WHITELISTED_PATH
                     || (parentMatchStatus == ScanSpecPathMatch.AT_WHITELISTED_CLASS_PACKAGE
-                            && scanSpec.classfileIsSpecificallyWhitelisted(relativePath))) {
+                            && scanSpec.classfileIsSpecificallyWhitelisted(relativePath))
+                    || (scanSpec.enableClassInfo && relativePath.equals("module-info.class"))) {
                 // Resource is whitelisted
                 addWhitelistedResource(resource, parentMatchStatus, subLog);
             }
