diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/hudson/tasks/junit/CaseResult.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/hudson/tasks/junit/CaseResult.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/hudson/tasks/junit/CaseResult.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/hudson/tasks/junit/CaseResult.java
@@ -270,7 +270,7 @@
     public String getTransformedTestName() {
-        return TestNameTransformer.getTransformedName(testName);
+        return TestNameTransformer.getTransformedName(getName());
     }
 
     public String getDisplayName() {
@@ -300,6 +300,9 @@
      */
     @Exported(visibility=999)
     public @Override String getName() {
+    	if (testName.isEmpty()) {
+    		return "(?)";
+    	}
         return testName;
     }
