diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/hudson/tasks/junit/CaseResult.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/hudson/tasks/junit/CaseResult.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/hudson/tasks/junit/CaseResult.java	2024-09-08 01:36:08.101766481 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/hudson/tasks/junit/CaseResult.java	2024-09-08 01:36:08.077766314 +1000
@@ -270,7 +270,7 @@
     }
 
     public String getTransformedTestName() {
-        return TestNameTransformer.getTransformedName(testName);
+        return TestNameTransformer.getTransformedName(getName());
     }
 
     public String getDisplayName() {
@@ -300,6 +300,9 @@
      */
     @Exported(visibility=999)
     public @Override String getName() {
+    	if (testName.isEmpty()) {
+    		return "(?)";
+    	}
         return testName;
     }
 
diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/hudson/tasks/junit/TestResultTest.java ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/hudson/tasks/junit/TestResultTest.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/hudson/tasks/junit/TestResultTest.java	2024-09-08 01:36:08.109766538 +1000
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/hudson/tasks/junit/TestResultTest.java	2024-09-08 01:36:08.085766369 +1000
@@ -246,6 +246,27 @@
         assertEquals("Wrong duration for test class", 93.0, class2.getDuration(), 0.1);
     }
 
+    @Issue("JENKINS-63113")
+    @Test
+    public void testTestcaseWithEmptyName() throws Exception {
+        TestResult testResult = new TestResult();
+        testResult.parse(getDataFile("junit-report-empty-testcasename.xml"));
+        testResult.tally();
+
+        assertEquals("Wrong number of testsuites", 1, testResult.getSuites().size());
+        assertEquals("Wrong number of test cases", 1, testResult.getTotalCount());
+
+        SuiteResult suite = testResult.getSuite("test.TestJUnit5FailingInBeforeAll");
+        assertNotNull(suite);
+
+        assertEquals("Wrong number of test classes", 1, suite.getClassNames().size());
+        CaseResult case1 = suite.getCases().get(0);
+
+        assertEquals("test.TestJUnit5FailingInBeforeAll.(?)", case1.getFullName());
+        assertEquals("(?)", case1.getDisplayName());
+        assertEquals("(?)", case1.getName());
+    }
+
     private static final XStream XSTREAM = new XStream2();
 
     static {
Only in ./RegMiner4APR-Regression-Bugs/BIC/src/test/resources/hudson/tasks/junit: junit-report-empty-testcasename.xml
