diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/crawlercommons/domains/EffectiveTldFinder.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/crawlercommons/domains/EffectiveTldFinder.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/crawlercommons/domains/EffectiveTldFinder.java	2025-12-11 23:35:06.000000000 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/crawlercommons/domains/EffectiveTldFinder.java	2025-12-11 23:34:54.000000000 +1100
@@ -82,12 +82,14 @@
  */
 public class EffectiveTldFinder {
     private static final Logger LOGGER = LoggerFactory.getLogger(EffectiveTldFinder.class);
+    
     public static final String ETLD_DATA = "/effective_tld_names.dat";
     public static final String COMMENT = "//";
     public static final String DOT_REGEX = "\\.";
     public static final String EXCEPTION = "!";
     public static final String WILD_CARD = "*.";
     public static final char DOT = '.';
+    
     private static EffectiveTldFinder instance = null;
     private Map<String, EffectiveTLD> domains = null;
     private boolean configured = false;
@@ -222,7 +224,7 @@
 
     /**
      * This method uses the effective TLD to determine which component of a FQDN
-     * is the NIC-assigned domain name.
+     * is the NIC-assigned domain name (aka "Paid Level Domain").
      * 
      * @param hostname
      *            a string for which to obtain a NIC-assigned domain name
@@ -235,7 +237,7 @@
 
     /**
      * This method uses the effective TLD to determine which component of a FQDN
-     * is the NIC-assigned domain name.
+     * is the NIC-assigned domain name (aka "Paid Level Domain").
      * 
      * @param hostname
      *            a string for which to obtain a NIC-assigned domain name
@@ -308,13 +310,14 @@
         public EffectiveTLD(String line) {
             if (line.startsWith(EXCEPTION)) {
                 exception = true;
-                domain = line.substring(1, line.length());
+                domain = line.substring(EXCEPTION.length(), line.length());
             } else if (line.startsWith(WILD_CARD)) {
                 wild = true;
-                domain = line.substring(2, line.length());
+                domain = line.substring(WILD_CARD.length(), line.length());
             } else {
                 domain = line;
             }
+            
             domain = normalizeName(domain);
         }
 
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/crawlercommons/domains/PaidLevelDomain.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/crawlercommons/domains/PaidLevelDomain.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/crawlercommons/domains/PaidLevelDomain.java	2025-12-11 23:35:06.000000000 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/crawlercommons/domains/PaidLevelDomain.java	2025-12-11 23:34:54.000000000 +1100
@@ -17,10 +17,6 @@
 package crawlercommons.domains;
 
 import java.net.URL;
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.Locale;
-import java.util.Set;
 import java.util.regex.Pattern;
 
 import org.slf4j.Logger;
@@ -35,19 +31,7 @@
 public class PaidLevelDomain {
     private static final Logger LOGGER = LoggerFactory.getLogger(PaidLevelDomain.class);
 
-    private static final String CC_TLDS = "ac ad ae af ag ai al am an ao aq ar as at au aw ax az ba bb bd be bf bg bh bi "
-                    + "bj bl bm bn bo br bs bt bv bw by bz ca cc cd cf cg ch ci ck cl cm cn co cr cu " + "cv cx cy cz de dj dk dm do dz ec ee eg eh er es et eu fi fj fk fm fo fr ga gb "
-                    + "gd ge gf gg gh gi gl gm gn gp gq gr gs gt gu gw gy hk hm hn hr ht hu id ie il " + "im in io iq ir is it je jm jo jp ke kg kh ki km kn kp kr kw ky kz la lb lc li "
-                    + "lk lr ls lt lu lv ly ma mc md me mf mg mh mk ml mm mn mo mp mq mr ms mt mu mv " + "mw mx my mz na nc ne nf ng ni nl no np nr nu nz om pa pe pf pg ph pk pl pm pn "
-                    + "pr ps pt pw py qa re ro rs ru rw sa sb sc sd se sg sh si sj sk sl sm sn so sr " + "st su sv sy sz tc td tf tg th tj tk tl tm tn to tp tr tt tv tw tz ua ug uk um "
-                    + "us uy uz va vc ve vg vi vn vu wf ws ye yt yu za zm zw";
-
-    private static final String G_TLDS = "aero arpa asia biz cat com coop edu gov info int jobs me mil mobi museum name net " + "org pro rs tel";
-
-    private static final Set<String> ccTLDs = new HashSet<String>(Arrays.asList(CC_TLDS.split(" ")));
-    private static final Set<String> gTLDs = new HashSet<String>(Arrays.asList(G_TLDS.split(" ")));
-
-    private static final Pattern IPV4_ADDRESS_PATTERN = Pattern.compile("(?:\\d{1,3}\\.){3}\\d{1,3}");
+    private static final Pattern IPV4_ADDRESS_PATTERN = Pattern.compile("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}");
 
     /**
      * Extract the PLD (paid-level domain) from the hostname. If the format
@@ -64,60 +48,20 @@
             return hostname;
         }
 
-        String[] subNames = hostname.split("\\.");
-        int numPieces = subNames.length;
-        if (numPieces <= 2) {
-            return hostname;
-        }
-
         // Check for ddd.ddd.ddd.ddd IPv4 format
-        if ((numPieces == 4) && (IPV4_ADDRESS_PATTERN.matcher(hostname).matches())) {
+        if (IPV4_ADDRESS_PATTERN.matcher(hostname).matches()) {
             return hostname;
         }
 
-        int firstHostPiece = 0;
-        if (ccTLDs.contains(subNames[numPieces - 1].toLowerCase(Locale.ROOT))) {
-            // We have a country code at the end. See if the preceding piece is
-            // either
-            // a two-letter name (country code or funky short gTLD), or one of
-            // the
-            // "well-known" gTLDs.
-            if (subNames[numPieces - 2].length() <= 2) {
-                // Must be xxx.co.jp format
-                firstHostPiece = numPieces - 3;
-            } else if (gTLDs.contains(subNames[numPieces - 2].toLowerCase(Locale.ROOT))) {
-                // Must be xxx.com.mx format
-                firstHostPiece = numPieces - 3;
-            } else {
-                // Must be xxx.it format
-                firstHostPiece = numPieces - 2;
-            }
-        } else if (gTLDs.contains(subNames[numPieces - 1].toLowerCase(Locale.ROOT))) {
-            if (ccTLDs.contains(subNames[numPieces - 2].toLowerCase(Locale.ROOT))) {
-                // Must be xxx.de.com format
-                firstHostPiece = numPieces - 3;
-            } else {
-                // Must be xxx.com format
-                firstHostPiece = numPieces - 2;
-            }
+        // Now use support in EffectiveTldFinder
+        String result = EffectiveTldFinder.getAssignedDomain(hostname, true);
+        if (result == null) {
+        	LOGGER.debug(String.format("Hostname %s isn't a valid FQDN", hostname));
+        	return hostname;
         } else {
-            LOGGER.debug("Unknown format for hostname: " + hostname);
-        }
-
-        if (firstHostPiece == 0) {
-            return hostname;
-        } else {
-            // Build the result from the firstHostPiece to numPices pieces.
-            StringBuilder result = new StringBuilder();
-            for (int i = firstHostPiece; i < numPieces; i++) {
-                result.append(subNames[i]);
-                result.append('.');
-            }
-
-            // Trim off final '.'
-            return result.deleteCharAt(result.length() - 1).toString();
+        	return result;
         }
-    } // getPLD
+    }
 
     /**
      * Extract the PLD (paid-level domain) from the URL.
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/crawlercommons/domains/PaidLevelDomainTest.java ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/crawlercommons/domains/PaidLevelDomainTest.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/test/java/crawlercommons/domains/PaidLevelDomainTest.java	2025-12-12 00:10:50.383285270 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/test/java/crawlercommons/domains/PaidLevelDomainTest.java	2025-12-12 00:17:27.071515546 +1100
@@ -84,14 +84,14 @@
 
     // Typical international domains look like xxx.it. So xxx.com.it is
     // actually the xxx subdomain for com.it
-    @Ignore @Test
+    @Test
     public final void testItalianDomains() {
         assertEquals("xxx.it", PaidLevelDomain.getPLD("xxx.it"));
         assertEquals("xxx.it", PaidLevelDomain.getPLD("www.xxx.it"));
         assertEquals("com.it", PaidLevelDomain.getPLD("xxx.com.it"));
     }
 
-    @Ignore @Test
+    @Test
     public final void testFinnishDomains() {
         assertEquals("fi.com", PaidLevelDomain.getPLD("www.fi.com"));
     }
