diff -r -u ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/apache/commons/io/input/BufferedFileChannelInputStream.java ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/apache/commons/io/input/BufferedFileChannelInputStream.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/src/main/java/org/apache/commons/io/input/BufferedFileChannelInputStream.java
+++ ./RegMiner4APR-Regression-Bugs/BIC/src/main/java/org/apache/commons/io/input/BufferedFileChannelInputStream.java
@@ -169,6 +169,9 @@
 
     @Override
     public synchronized int available() throws IOException {
+        if (!refill()) {
+            return EOF;
+        }
         return byteBuffer.remaining();
     }
 
